File: src/types/squad.ts
/**
 * Frontend-specific types for the squad system.
 * Extends the base model types with UI-specific fields.
 */

export interface SquadListItem {
  id: string;
  name: string;
  description: string | null;
  avatar_url: string | null;
  is_personal: boolean;
  member_count: number;
  current_user_role: string | null;
  created_at: string;
}

export interface SquadDetail {
  id: string;
  name: string;
  description: string | null;
  avatar_url: string | null;
  invite_code: string;
  is_personal: boolean;
  max_members: number;
  member_count: number;
  created_by: string;
  created_at: string;
  updated_at: string;
  current_user_role: string | null;
}

export interface MemberProfile {
  id: string;
  email: string;
  display_name: string | null;
  full_name: string | null;
  avatar_url: string | null;
}

export interface SquadMemberItem {
  id: string;
  squad_id: string;
  user_id: string;
  role: string;
  joined_at: string;
  profile: MemberProfile | null;
}

export interface InvitationDetail {
  id: string;
  squad_id: string;
  squad_name: string | null;
  invited_by: string;
  inviter_name: string | null;
  token: string;
  role: string;
  status: string;
  expires_at: string;
  created_at: string;
}

export interface InvitationPublic {
  squad_name: string;
  squad_avatar_url: string | null;
  squad_member_count: number;
  role: string;
  status: string;
  expires_at: string;
  inviter_name: string | null;
  is_expired: boolean;
}

export interface SidebarUser {
  id: string;
  email: string;
  name: string;
  avatar_url: string | null;
}

File: backend/app/models/squad.py
"""SQLAlchemy models: Squad, SquadMember, Invitation."""

from datetime import datetime
from uuid import UUID, uuid4

from sqlalchemy import Boolean, DateTime, ForeignKey, Integer, String, Text, func
from sqlalchemy.dialects.postgresql import JSONB, UUID as PGUUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.db.session import Base


class Squad(Base):
    __tablename__ = "squads"

    id: Mapped[UUID] = mapped_column(PGUUID(as_uuid=True), primary_key=True, default=uuid4)
    name: Mapped[str] = mapped_column(String(200), nullable=False)
    description: Mapped[str | None] = mapped_column(Text)
    avatar_url: Mapped[str | None] = mapped_column(Text)
    invite_code: Mapped[str] = mapped_column(String(20), nullable=False, unique=True)
    is_personal: Mapped[bool] = mapped_column(Boolean, default=False, nullable=False)
    max_members: Mapped[int] = mapped_column(Integer, default=50, nullable=False)
    member_count: Mapped[int] = mapped_column(Integer, default=0, nullable=False, server_default="0")
    created_by: Mapped[UUID] = mapped_column(
        PGUUID(as_uuid=True),
        ForeignKey("auth.users.id", ondelete="RESTRICT"),
        nullable=False,
    )
    metadata_: Mapped[dict] = mapped_column("metadata", JSONB, default=dict, nullable=False)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now())
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), onupdate=func.now()
    )
    deleted_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))

    members: Mapped[list["SquadMember"]] = relationship(back_populates="squad")
    vaults: Mapped[list] = relationship("Vault", back_populates="squad")


class SquadMember(Base):
    __tablename__ = "squad_members"

    id: Mapped[UUID] = mapped_column(PGUUID(as_uuid=True), primary_key=True, default=uuid4)
    squad_id: Mapped[UUID] = mapped_column(
        PGUUID(as_uuid=True), ForeignKey("squads.id", ondelete="CASCADE"), nullable=False
    )
    user_id: Mapped[UUID] = mapped_column(
        PGUUID(as_uuid=True),
        ForeignKey("auth.users.id", ondelete="CASCADE"),
        nullable=False,
    )
    role: Mapped[str] = mapped_column(String(30), default="member", nullable=False)
    joined_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now())
    removed_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))

    squad: Mapped["Squad"] = relationship(back_populates="members")
    # user_id and profiles.id both reference auth.users.id independently (no
    # direct FK between squad_members and profiles), so the join must be
    # spelled out explicitly.
    profile: Mapped["Profile"] = relationship(
        "Profile",
        primaryjoin="SquadMember.user_id == Profile.id",
        foreign_keys="[SquadMember.user_id]",
        viewonly=True,
        lazy="joined",
    )


class Invitation(Base):
    __tablename__ = "invitations"

    id: Mapped[UUID] = mapped_column(PGUUID(as_uuid=True), primary_key=True, default=uuid4)
    squad_id: Mapped[UUID] = mapped_column(
        PGUUID(as_uuid=True), ForeignKey("squads.id", ondelete="CASCADE"), nullable=False
    )
    invited_by: Mapped[UUID] = mapped_column(
        PGUUID(as_uuid=True),
        ForeignKey("auth.users.id", ondelete="CASCADE"),
        nullable=False,
    )
    invited_email: Mapped[str | None] = mapped_column(String(320))
    invited_user_id: Mapped[UUID | None] = mapped_column(
        PGUUID(as_uuid=True), ForeignKey("auth.users.id", ondelete="CASCADE")
    )
    token: Mapped[str] = mapped_column(String(64), nullable=False, unique=True)
    role: Mapped[str] = mapped_column(String(30), default="member", nullable=False)
    status: Mapped[str] = mapped_column(String(20), default="pending", nullable=False)
    expires_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False)
    responded_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))
    accepted_by: Mapped[UUID | None] = mapped_column(
        PGUUID(as_uuid=True), ForeignKey("auth.users.id", ondelete="SET NULL")
    )
    accepted_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now())

    squad: Mapped["Squad"] = relationship(foreign_keys=[squad_id], lazy="joined")
